Merge two Linux APIs for mapping kernel modules to pointers. Fix bugs…

atcuno opened the pull request:
… in new API. Add deprecation warnings throughout old API.

This PR fixes a serious set of issues in the Linux APIs related to mapping kernel pointers to their module and symbol name (a critical aspect of rootkit detection). Previously, I had written an API in the `symbols/linux/__init__.py` that performed these tasks and were very easy for plugin writers to use - you called a function to get the handler information then you called a second function anytime you needed to map an address inside the handlers.

There was a recent push for covering the tracing APIs (ftrace and tracepoints) that added an entirely new API for mapping modules and symbols, but this work had three major issues that were not handled in attempting to replicate my old API, but with new features added:

1) This API did not incorporate the kernel .text space in its "modules" set, which meant that every pointer inside the kernel was reported as a dash (rootkit marker) in plugins. In the mass testing runs, this led to every single sample reporting rootkits all over the place.

2) This API did not directly incorporate the symbol lookups, leading to a bunch of copy/paste code between plugins that wanted to use it - see the diffs in this PR for ftrace and tracepoint - now they match the easy call setup from my old API, but also incorporates the hidden module searching.

3) The PRs for the new API left the old API in place, plus many existing plugins using the old API. This left Volatility with two public APIs for mapping Linux kernel pointers, neither of which were complete. The updated API is unified and gets the best features of each (hidden modules, mapping inside kernel .text), while leaving it very easy on plugin writers.

This PR fixes all the above issues, with the minimal amount of code possible. To rectify all the issues, and get it to where all the plugins will use new API, I did the following:

1) Deprecated all the old code paths

2) Bumped the Linux utilities version on the plugins using the old paths, but left them still using the deprecated paths to minimze the size of this PR. In a PR right after this one is merged, I will update them to use the new paths to avoid the deprecation warnings.

3) Volatility 3 has the really nice `deprecated_method` feature, but this does not cover when an entire function will be deprecated and going away without a direct replacement. I added this feature. This feature was necessary as the old API functions that worked with the tuples of (name, start, end) are just in no way compatible with the new API.

4) Moved all shared code to the utility class to avoid plugins importing from multiple plugins + the utility. I assume these functions were spread out through plugins before the versioned Modules utility, but that is the wrong way to approach it now. This also prevents and avoids the need for circular imports.

5) I fixed up typing where I noticed it in plugins already getting a major version bump.

atcuno committed:
Merge two Linux APIs for mapping kernel modules to pointers. Fix bugs in new API. Add deprecation warnings throughout old API.

atcuno committed:
update for black

ikelos commented on volatility3/framework/__init__.py:
```diff
@@ -69,6 +69,22 @@ def require_interface_version(*args) -> None:
+            @functools.wraps(deprecated_func)
+            def wrapper(*args, **kwargs):
+                warnings.warn(
+                    f"This API ({deprecated_func.__module__}.{deprecated_func.__qualname__}) will be removed in a release very soon. {message}",
```
Might be worth adding an extra parameter with a version number to say when it'll definitely be deprecated by?  Problem is the versions bump at unexpected times, so perhaps a datetime for removal is better?  It would also make it easy for us to grep them all at release and remove the ones whose time has come?

ikelos commented on volatility3/framework/symbols/linux/__init__.py:
```diff
@@ -367,9 +368,8 @@ def files_descriptors_for_process(
-        replacement=linux_utilities_modules.Modules.mask_mods_list,
-        replacement_version=(1, 0, 0),
+    @Deprecation.method_being_removed(
+        "This method is not needed. The correct API for mapping kernel pointers to modules is `linux_utilities_modules.Modules.run_module_scanners`"
```
The message should be something a little softer but more direct.  "This method is no longer recommended and will be removed from the framework in a future release.  Code using this method should be adapted to make use of `linux_utilities_modules.Modules.run_module_scanners`"

Not needed implies it's fine to keep using it, and "correct API" sounds like you're calling this one wrong, which is wasn't right up until we deprecated it (and it'll actually still work afterwards just fine for a while).

atcuno replied:
ok, `method_being_removed` already adds the generic deprecation method so will shorten to just the replacement method

ikelos commented on volatility3/framework/plugins/linux/hidden_modules.py:
```diff
@@ -3,7 +3,8 @@
 import logging
 from typing import List, Set, Tuple, Iterable
-from volatility3.framework import renderers, interfaces, exceptions, objects
+import volatility3.framework.symbols.linux.utilities.modules as linux_utilities_modules
```
```suggestion
from volatility3.framework.symbols.linux.utilities import modules as linux_utilities_modules
```
Just a stylistic point and has no impact on the code, but I kinda think it's easier to stick with the "always import from" and then do the renaming after?

ikelos commented on volatility3/framework/plugins/linux/hidden_modules.py:
```diff
@@ -3,7 +3,8 @@
 from typing import List, Set, Tuple, Iterable
-from volatility3.framework import renderers, interfaces, exceptions, objects
+import volatility3.framework.symbols.linux.utilities.modules as linux_utilities_modules
+from volatility3.framework import renderers, interfaces, exceptions, Deprecation
```
```suggestion
from volatility3 import framework
from volatility3.framework import renderers, interfaces, exceptions
```

Then use `framework.Deprecation`?  I'd expect David's future tool to flag this, so... might as well do it right first time round?

ikelos replied:
To be fair, it's a little odd they're in a class anyway?  The class isn't versioned or anything, just seems to be used for grouping.  I'd almost prefer a `deprecation.py` and have them just live straight in there?

atcuno replied:
Yea, the Deprecation is only previously used in `linux/symbols/__init__.py` so far, which is being changed in this PR anyway. I thought the capital D was odd so I can fix that. Also, can put into its own file under framework/. Your other feedback is good too. I like the date idea, so I will do that tomorrow as well. Brain is too burned for today, but will knock out first thing in the morning.

ikelos commented on volatility3/framework/plugins/linux/hidden_modules.py:
```diff
@@ -29,46 +30,30 @@ def get_requirements(cls) -> List[interfaces.configuration.RequirementInterface]
 
-    @classmethod
+    @staticmethod
+    @Deprecation.deprecated_method(
```
Yep, shouldn't start with a capital D, definitely prefer these methods moving to a module of their own please.

ikelos replied:
Happy for you to do it in this PR, but fine if you just want to put a bug number here and we can do it later, but we might as well do it before use becomes widespread.

ikelos commented on volatility3/framework/plugins/linux/hidden_modules.py:
```diff
@@ -29,46 +30,30 @@ def get_requirements(cls) -> List[interfaces.configuration.RequirementInterface]
+        replacement_version=(2, 0, 0),
+    )
     @classmethod
     def _get_module_address_alignment(
```
Surprised I let a private method be exposed as a classmethod, but just as well to deprecate it, thanks!

ikelos replied:
(It probably had to be a classmethod to be called from within another classmethod, that makes sense).

ikelos commented on volatility3/framework/plugins/linux/lsmod.py:
```diff
@@ -6,7 +6,8 @@
 from typing import List, Iterable
 
-from volatility3.framework import exceptions, renderers, constants, interfaces
+import volatility3.framework.symbols.linux.utilities.modules as linux_utilities_modules
+from volatility3.framework import exceptions, renderers, interfaces, Deprecation
```
```suggestion
from volatility3.framework.symbols.linux.utilities import modules as linux_utilities_modules
from volatility3 import framework
from volatility3.framework import exceptions, renderers, interfaces
```

ikelos commented on volatility3/framework/plugins/linux/lsmod.py:
```diff
@@ -29,43 +30,38 @@ def get_requirements(cls) -> List[interfaces.configuration.RequirementInterface]
         ]
 
     @classmethod
+    @Deprecation.deprecated_method(
```
```suggestion
    @framework.Deprecation.deprecated_method(
```

ikelos commented on volatility3/framework/plugins/linux/hidden_modules.py:
```diff
@@ -88,27 +73,14 @@ def _get_module_address_alignment(
+            context, vmlinux_module_name
+        )
 
+    @Deprecation.deprecated_method(
```
```suggestion
    @framework.Deprecation.deprecated_method(
```

ikelos commented on volatility3/framework/plugins/linux/hidden_modules.py:
```diff
@@ -29,46 +30,30 @@ def get_requirements(cls) -> List[interfaces.configuration.RequirementInterface]
 
-    @classmethod
+    @staticmethod
+    @Deprecation.deprecated_method(
```
```suggestion
    @framework.Deprecation.deprecated_method(
```

ikelos commented on volatility3/framework/plugins/linux/check_modules.py:
```diff
@@ -29,58 +30,33 @@ def get_requirements(cls) -> List[interfaces.configuration.RequirementInterface]
         ]
 
     @classmethod
+    @Deprecation.deprecated_method(
```
```suggestion
    @framework.Deprecation.deprecated_method(
```

ikelos commented on volatility3/framework/plugins/linux/check_modules.py:
```diff
@@ -3,22 +3,23 @@
+from typing import List, Dict
 
-from volatility3.framework import interfaces, renderers, exceptions, constants
+import volatility3.framework.symbols.linux.utilities.modules as linux_utilities_modules
+from volatility3.framework import interfaces, renderers, Deprecation
```
```suggestion
from volatility3.framework.symbols.linux.utilities import modules as linux_utilities_modules
from volatility3 import framework
from volatility3.framework import interfaces, renderers
```

ikelos commented on volatility3/framework/plugins/linux/modxview.py:
```diff
@@ -3,8 +3,10 @@
-from volatility3.plugins.linux import lsmod, check_modules, hidden_modules
-from volatility3.framework import interfaces
+
+import volatility3.framework.symbols.linux.utilities.modules as linux_utilities_modules
+
+from volatility3.framework import interfaces, Deprecation
```
```suggestion
from volatility3.framework.symbols.linux.utilities import modules as linux_utilities_modules
from volatility3 import framework
from volatility3.framework import interfaces
```

ikelos commented on volatility3/framework/plugins/linux/modxview.py:
```diff
@@ -54,6 +48,10 @@ def get_requirements(cls) -> List[interfaces.configuration.RequirementInterface]
         ]
 
     @classmethod
+    @Deprecation.deprecated_method(
```
```suggestion
    @framework.Deprecation.deprecated_method(
```

ikelos commented on volatility3/framework/plugins/linux/modxview.py:
```diff
@@ -67,83 +65,53 @@ def flatten_run_modules_results(
+        )
 
     @classmethod
+    @Deprecation.deprecated_method(
```
```suggestion
    @framework.Deprecation.deprecated_method(
```

ikelos commented on volatility3/framework/plugins/linux/tracing/ftrace.py:
```diff
@@ -10,11 +10,9 @@
 from dataclasses import dataclass
 
 import volatility3.framework.symbols.linux.utilities.modules as linux_utilities_modules
```
```suggestion
from volatility3.framework.symbols.linux.utilities import modules as linux_utilities_modules
```

ikelos commented on volatility3/framework/plugins/linux/tracing/ftrace.py:
```diff
@@ -264,16 +217,15 @@ def _generator(self):
-        known_modules = modxview.Modxview.run_modules_scanners(
-            self.context, kernel_name, run_hidden_modules=False
+        known_modules = linux_utilities_modules.Modules.run_modules_scanners(
+            self.context, kernel_name, run_hidden_modules=True
```
Is hidden_modules fast enough to include by default now?

ikelos commented on volatility3/framework/symbols/linux/__init__.py:
```diff
@@ -367,9 +368,8 @@ def files_descriptors_for_process(
-    @Deprecation.deprecated_method(
-        replacement=linux_utilities_modules.Modules.mask_mods_list,
-        replacement_version=(1, 0, 0),
+    @Deprecation.method_being_removed(
```
```suggestion
    @framework.Deprecation.method_being_removed(
```

ikelos commented on volatility3/framework/symbols/linux/__init__.py:
```diff
@@ -385,13 +385,18 @@ def mask_mods_list(
         return linux_utilities_modules.Modules.mask_mods_list(context, layer_name, mods)
 
     @classmethod
+    @Deprecation.method_being_removed(
```
```suggestion
    @framework.Deprecation.method_being_removed(
```

ikelos commented on volatility3/framework/symbols/linux/utilities/modules.py:
```diff
@@ -1,77 +1,121 @@
 
 from volatility3 import framework
-from volatility3.framework import constants, interfaces
+from volatility3.framework import (
+    constants,
+    interfaces,
+    Deprecation,
+    exceptions,
+    objects,
+)
```
```suggestion
from volatility3.framework import (
    constants,
    interfaces,
    exceptions,
    objects,
)
```

ikelos commented on volatility3/framework/symbols/linux/utilities/modules.py:
```diff
@@ -1,77 +1,121 @@
+        return None, None
 
     @classmethod
+    @Deprecation.method_being_removed(
```
```suggestion
    @framework.Deprecation.method_being_removed(
```

ikelos commented on volatility3/framework/symbols/linux/utilities/modules.py:
```diff
@@ -1,77 +1,121 @@
 
     @classmethod
+    @Deprecation.method_being_removed(
+        "This method is not needed. The correct API for mapping kernel pointers to modules is `linux_utilities_modules.Modules.run_module_scanners`"
```
Again, just soften the "correct" API, and we don't need to say it's not needed, just that it's being removed.

ikelos commented on volatility3/framework/symbols/linux/utilities/modules.py:
```diff
@@ -1,77 +1,121 @@
-    ) -> List[Tuple[str, int, int]]:
+        kernel_layer_name: str,
+        mods: Iterator[extensions.module],
+    ) -> List[extensions.module]:
```
Errr, the ocde looks like it still returns tuples?  It's fine because it's deprecated but... I think the original typing information was correct, at least for the output?

atcuno replied:
Good catch.

ikelos commented on volatility3/framework/symbols/linux/utilities/modules.py:
```diff
@@ -116,3 +163,369 @@ def lookup_module_address(
+        )
+
+    @classmethod
+    def run_modules_scanners(
```
Something to consider here is whether we want to do something like the linux pslist thing we did?  With different methods with the same signature that scanned different ways?  Then you just hand in the list of scanners you want rather than boolean flags to turn on and off specific sets of scanners?

atcuno replied:
I was going to revisit the `implementation_method` for this after this PR goes on. I honestly don't even like the skipping hidden part, but left it here as it was there already. We should just get all the sources all the time.

Abyss-W4tcher replied:
The issue with doing the hidden_modules scanning by default, is that if the caller never needs it (i.e. it is able to resolve all symbols to modules without it), then we will just have wasted time and CPU scanning the space. 

The previous design was triggering the scanning on a per-need basis, if a resolution issue was raised. I think we should leave the choice, depending on the use case ?

This is even more true on unaligned spaces, where hidden_modules can take quite a long time running.

atcuno replied:
Its only slow when the path for 1 byte alignment is hit, which I am going to fix in a follow up PR. Otherwise, it didn't change speed much in my testing since it only scans inside the vmmemap bounds anyway. If there are samples where this is problematic let me know though as I am confident we can make the scanner quick vs having to conditionally run it.

ikelos replied:
Can you raise a ticket and note it here to make sure this gets addressed in the future please?  (The redo to make the set of gatherers more generic/pluggable)

atcuno replied:
The ticket is here now: https://github.com/volatilityfoundation/volatility3/issues/1699

ikelos reviewed (commented):
Cool, looks really good!  For a minute I was worried we were changing the `LinuxUtilities` MAJOR version, but then saw it was just the Modules extension.  5:)

There's a couple considerations about changing the API but they may be excessive, so feel free to reject them.  I really don't like the way there's a Deprecation class and we're importing it directly, I'd really like that cleaned up as part of this PR please but will accept an open bug if you want to put it off a bit.  Otherwise looks really good and simplifies/collects together a lot of the functionality, nice!  5:D

atcuno committed:
Update volatility3/framework/plugins/linux/hidden_modules.py

Co-authored-by: ikelos <ikelos@users.noreply.github.com>

ikelos commented on volatility3/framework/plugins/linux/hidden_modules.py:
```diff
@@ -3,7 +3,8 @@
 import logging
 from typing import List, Set, Tuple, Iterable
-from volatility3.framework import renderers, interfaces, exceptions, objects
+from volatility3.framework.symbols.linux.utilities import modules as linux_utilities_modules
```
```suggestion
from volatility3.framework.symbols.linux.utilities import (
    modules as linux_utilities_modules,
)
```
Black complains about it

ikelos committed:
Update volatility3/framework/plugins/linux/hidden_modules.py

Ensure black doesn't complain

atcuno commented:
@ikelos I addressed all your feedback except for adding a datetime to deprecation headings as I am not sure what the implementation would look like.

So, if the calls took a `datetime`, what would happen with them? The specific date printed in the warning message? Backtrace the calling plugin with a new deprecation exception? Or would they be a parameter just to be grepped before releases?

ikelos commented:
Yeah, didn't need to be a `datetime` per-se, even a date string would do (`YYYY-MM-DD` please), so that we could easily search for ones that should be removed, and we could tell everyone exactly when the method should no longer be used...

ikelos commented:
We could theoretically throw a full on exception if it tries to be used after that date, but I think we should just allow it until we get around to removing the method (hopefully it'll happen ever release, one for us to check during the release process).

ikelos commented:
We need to come up with a general policy, I'd suggest at least one year (although I'd possibly go as low as 6 months, because with our release cadence it's more likely to be 9 months) or when a MAJOR version bump happens (no point carry over cruft when everything changes)...

atcuno committed:
Deprecation API updates

atcuno committed:
Bring back change after merge conflict

atcuno committed:
Add new deprecation module

atcuno committed:
Add removal_date to deprecation API

atcuno committed:
Add removal_date to deprecation API

atcuno commented:
I did 6 months plus a few days @ikelos

ikelos commented on volatility3/framework/symbols/linux/utilities/modules.py:
```diff
@@ -1,77 +1,122 @@
+
+        kernel_layer = context.layers[kernel.layer_name]
+
+        if modules[0].start != modules[0].start & kernel_layer.address_mask:
```
This seems a *really* suspect test, it's checking that high bits aren't set?  Why is it only doing it for one entry?  What if the list is empty (for some reason)?

atcuno replied:
I will add the check for an empty list (good catch), but I want to keep the index 0 check as it will enforce callers going through the right API flow to reach that code. All kernel modules will have the high bits set initially, so if the right API flow isn't used then non-masked ones will make it there and everything processes incorrectly.

ikelos replied:
Shouldn't it be something like `any(modules[i].start =! modules[i].start & kernel_layer.address_mask for i in modules)` though?  Picking out the first one seems like a half hearted check?

atcuno replied:
If the caller comes through the right API then all will be masked the same, if they don't then none will be masked... if they somehow randomly enumerate on their own and only mask half.. then they are just trying to be broken.

ikelos replied:
That's not a good argument for why not to check them all?  You're applying trust to something you could check, so why not just check it?  Otherwise I'm certain someone could write code that masks the first value and then makes this call.  There's little point in half a test (and it also requires that the list has elements, whereas any/all would not)...

atcuno replied:
Updated

ikelos commented on volatility3/framework/symbols/linux/utilities/modules.py:
```diff
@@ -1,77 +1,122 @@
+        if len(matches) >= 1:
+            if len(matches) > 1:
+                warnings.warn(
+                    f"Address {hex(target_address)} fits in modules at {[hex(module.start) for module in matches]}, indicating potential modules memory space overlap.",
```
Since this doesn't then bomb out, you should say which one will win.

atcuno replied:
added

ikelos commented on volatility3/framework/symbols/linux/utilities/modules.py:
```diff
@@ -116,3 +165,369 @@ def lookup_module_address(
+        start_addr = start_addr.vol.offset & mask
+
+        end_addr = kernel.object_from_symbol("_etext")
+        end_addr = end_addr.vol.offset & mask
```
```suggestion
        address_mask = context.layers[kernel.layer_name].address_mask

        start_addr = kernel.object_from_symbol("_text")
        start_addr = start_addr.vol.offset & address_mask

        end_addr = kernel.object_from_symbol("_etext")
        end_addr = end_addr.vol.offset & address_mask
```
Could we call this address_mask to keep it consistent with other methods please?

atcuno replied:
changed

ikelos commented on volatility3/framework/symbols/linux/utilities/modules.py:
```diff
@@ -116,3 +165,369 @@ def lookup_module_address(
+        if run_hidden_modules:
+            known_modules_addresses = set(
+                context.layers[kernel.layer_name].canonicalize(modinfo.start)
+                for modinfo in run_results["kernel"]
+                + run_results["lsmod"]
+                + run_results["check_modules"]
```
```suggestion
                for modinfo in cls.flatten_run_modules_results(run_results)
```
This should probably be driven by the contents of the list, rather than hard coding the available tests, which looks to be what flatten_run_modules_results does?

atcuno replied:
Yes this will be a part of: https://github.com/volatilityfoundation/volatility3/issues/1699

ikelos commented on volatility3/framework/symbols/linux/utilities/modules.py:
```diff
@@ -116,3 +165,369 @@ def lookup_module_address(
+
+        modules = vmlinux.object_from_symbol(symbol_name="modules").cast("list_head")
+
+        table_name = modules.vol.type_name.split(constants.BANG)[0]
```
This should just be `vmlinux.symbol_table_name`, no?

atcuno replied:
Good call. Much cleaner

ikelos reviewed (commented):
Looks good (assuming a follow up to make the types of module scans more... modular... so to speak).  A couple minor points but no show stoppers...

atcuno commented:
@ikelos I managed to break the commit history while rebasing. I asked @dgmcdona to look at it. Will finish it up after.

ikelos commented:
@j-t-1, please avoid modifying other people's pull requests.  It can cause a huge amount of confusion, and in general, it's better to have any changes you want to make to go through  a proper PR so they can be reviewed...

j-t-1 commented:
> @j-t-1, please avoid modifying other people's pull requests. It can cause a huge amount of confusion, and in general, it's better to have any changes you want to make to go through a proper PR so they can be reviewed...

Apologies for this. If I understand what has caused this, I deleted `pid_list = pid_list or []`, which I then realised broke the iteration in the line below, so I undeleted it in a subsequent commit, but only the first was used in a rebase. I will not deliberately modify other people's pull requests (if there is a way to remove access for me to do this, please do as one less potential error pathway).

ikelos commented:
Thanks, we will be improving permissions so that we don't have to spot when things go wrong, but can prevent them from being done in the first place.  I fully understand it was done intending to improve the commit, but we probably ought to prevent others from modifying someone's PR as you say to reduce potential error pathways.  Thanks for understanding...  5:)

ikelos commented on volatility3/framework/deprecation.py:
```diff
@@ -0,0 +1,81 @@
+from volatility3.framework.configuration import requirements
+
+
+def method_being_removed(message: str, removal_date: str):
```
Any chance of a docstring for this please?

atcuno replied:
added

ikelos reviewed (commented):
Looking really good, just the currently open comments to resolve please

atcuno commented:
all the feedback should be addressed now @ikelos

atcuno committed:
Address feedback

ikelos reviewed (approved):
Ok, looks good (although I'd still have preferred the `from ... import ... as ...` syntax, but it's not a showstopper)...  5;P
